incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -4289,54 +4407,23 @@ private void setInitialClosed()
 
                 if (MqttState.closed(state))
                 {
-                    streams.remove(topicKey);
-
-                    final MutableInteger activeStreams = activeStreamsByTopic.get(topicKey);
-
-                    assert activeStreams != null;
-
-                    activeStreams.value--;
-
-                    assert activeStreams.value >= 0;
-
-                    if (activeStreams.value == 0)
-                    {
-                        activeStreamsByTopic.remove(topicKey);
-                    }
+                    subscribeStreams.remove(clientKey);
                 }
             }
 
-            private void publishWillMessage(
-                long traceId)
-            {
-                final OctetsFW payload = willPayload;
-                final int payloadSize = payload.sizeof();
-
-                int reserved = payloadSize + initialPad;
-                boolean canPublish = initialSeq + reserved <= initialAck + initialMax;
-                int dataFlags = 0x03;
 
-                if (canPublish && debitorIndex != NO_DEBITOR_INDEX && reserved != 0)
-                {
-                    final int minimum = reserved; // TODO: fragmentation
-                    reserved = debitor.claim(debitorIndex, initialId, minimum, reserved);
+            private void setSubscribeReplyClosed()
+            {
+                assert !MqttState.replyClosed(state);
 
-                    if (reserved != minimum)
-                    {
-                        dataFlags &= ~0x01;
-                    }
-                }
+                state = MqttState.closeReply(state);
```

Comment:
The `state` updates should be updated in the `on...` and `do...` methods directly.